src/main/scala/mesosphere/marathon/core/storage/store/impl/zk/ZkPersistenceStore.scala

```diff
@@ -0,0 +1,261 @@
+package mesosphere.marathon.core.storage.store.impl.zk
+
+import java.time.OffsetDateTime
+import java.time.format.DateTimeFormatter
+import java.util.UUID
+
+import akka.actor.{ ActorRefFactory, Scheduler }
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+import akka.{ Done, NotUsed }
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.Protos.{ StorageVersion, ZKStoreEntry }
+import mesosphere.marathon.StoreCommandFailedException
+import mesosphere.marathon.core.storage.store.impl.{ BasePersistenceStore, CategorizedKey }
+import mesosphere.marathon.metrics.Metrics
+import mesosphere.marathon.storage.migration.Migration
+import mesosphere.marathon.util.{ Retry, Timeout, toRichFuture }
+import mesosphere.util.{ CapConcurrentExecutions, CapConcurrentExecutionsMetrics }
+import org.apache.zookeeper.KeeperException
+import org.apache.zookeeper.KeeperException.{ NoNodeException, NodeExistsException }
+import org.apache.zookeeper.data.Stat
+
+import scala.async.Async.{ async, await }
+import scala.collection.immutable.Seq
+import scala.concurrent.duration.Duration
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.control.NonFatal
+import scala.util.{ Failure, Success }
+
+case class ZkId(category: String, id: String, version: Option[OffsetDateTime]) {
+  private val bucket = math.abs(id.hashCode % ZkId.HashBucketSize)
+  def path: String = version.fold(f"/$category/$bucket%x/$id") { v =>
```

Comment:
why is `path` a `def` instead of a `val`? Couldn't this be declared as:
```
val path: String = {
  val bucket = math.abs(...)
  version.fold(...) { ... }
}
```